private/buf/cmd/buf/command/beta/studioagent/studioagent.go

```diff
@@ -0,0 +1,241 @@
+	)
+	flagSet.StringVar(
+		&f.CACert,
+		caCertFlagName,
+		"",
+		"The CA cert to be used in the client and server TLS configuration.",
+	)
+	flagSet.StringVar(
+		&f.ClientCert,
+		clientCertFlagName,
+		"",
+		"The cert to be used in the client TLS configuration.",
+	)
+	flagSet.StringVar(
+		&f.ClientKey,
+		clientKeyFlagName,
+		"",
+		"The key to be used in the client TLS configuration.",
+	)
+	flagSet.StringVar(
+		&f.ServerCert,
+		serverCertFlagName,
+		"",
+		"The cert to be used in the server TLS configuration.",
+	)
+	flagSet.StringVar(
+		&f.ServerKey,
+		serverKeyFlagName,
+		"",
+		"The key to be used in the server TLS configuration.",
+	)
+}
+
+func run(
+	ctx context.Context,
+	container appflag.Container,
+	flags *flags,
+) error {
+	// CA cert pool is optional. If it is nil, TLS uses the host's root CA set.
+	var caCertPool *x509.CertPool
+	var err error
+	if flags.CACert != "" {
+		caCertPool, err = newCACertPool(flags.CACert)
+		if err != nil {
+			return err
+		}
+	}
+	// client TLS config is optional. If it is nil, it uses the default configuration from http2.Transport.
+	var clientTLSConfig *tls.Config
+	if flags.ClientCert != "" || flags.ClientKey != "" {
+		clientTLSConfig, err = newTLSConfig(caCertPool, flags.ClientCert, flags.ClientKey)
+		if err != nil {
+			return fmt.Errorf("cannot create new client TLS config: %w", err)
+		}
+	}
+	// server TLS config is optional. If it is nil, we serve with a h2c handler.
+	var serverTLSConfig *tls.Config
+	if flags.ServerCert != "" || flags.ServerKey != "" {
+		serverTLSConfig, err = newTLSConfig(caCertPool, flags.ServerCert, flags.ServerKey)
+		if err != nil {
+			return fmt.Errorf("cannot create new server TLS config: %w", err)
+		}
+	}
+	mux := bufstudioagent.NewHandler(
+		container.Logger(),
+		container.Arg(0), // the origin from command argument
+		clientTLSConfig,
+		stringutil.SliceToMap(flags.DisallowedHeaders),
+		flags.ForwardHeaders,
+	)
+	if serverTLSConfig == nil {
+		mux = h2c.NewHandler(mux, &http2.Server{
+			IdleTimeout: defaultIdleTimeout,
+		})
+	}
+	httpServer := &http.Server{
+		Handler:           mux,
+		ReadHeaderTimeout: defaultReadHeaderTimeout,
+		TLSConfig:         serverTLSConfig,
+	}
+	var httpListenConfig net.ListenConfig
+	httpListener, err := httpListenConfig.Listen(ctx, "tcp", fmt.Sprintf("0.0.0.0:%s", flags.Port))
+	if err != nil {
+		return err
+	}
+	jobs := []func(context.Context) error{
+		func(ctx context.Context) error {
+			return httpServe(httpServer, httpListener)
+		},
+		func(ctx context.Context) error {
+			<-ctx.Done()
+			ctx, cancel := context.WithTimeout(context.Background(), defaultShutdownTimeout)
+			defer cancel()
+			return httpServer.Shutdown(ctx)
+		},
+	}
+	if err := thread.Parallelize(ctx, jobs); err != http.ErrServerClosed {
+		return err
+	}
+	return nil
+}
+
+func newCACertPool(caCertFile string) (*x509.CertPool, error) {
```

Comment:
Lets refactor this a bit to use our `certclient` package, export:

https://github.com/bufbuild/buf/blob/6092bfe2dfa0052644676c4397943133bdb94909/private/pkg/cert/certclient/util.go#L24-L25